File: docs/Befunde_und_Verbesserungen/prototypes/04-try-catch-definite-assignment/vergleich.kt
// 04 Vergleich: Kotlin — `try` ist ein Ausdruck; ein catch-Zweig, der mit `return` endet,
// hat den Typ Nothing und fällt aus der Unifikation heraus, also ist `opts` ein Options.
// (C# erreicht Stufe (a) ohne Ausdruck: Definite Assignment zählt die Body-Zuweisung,
//  wenn jeder catch die Methode verlässt — siehe vergleich.cs)

class UsageError(val what: String) : Exception("usage: $what")
data class Options(val limit: Int, val verbose: Boolean)

fun parseArgs(args: List<String>): Options {
    if (args.isEmpty()) throw UsageError("need at least one argument")
    return Options(args.size, args[0] == "-v")
}

fun runB(args: List<String>): Int {
    val opts = try { parseArgs(args) } catch (e: UsageError) { println(e.message); return 2 }
    println("B: limit ${opts.limit} verbose ${opts.verbose}")
    return 0
}

// "Grund egal": runCatching(...).getOrNull() entspricht Swift `try?` / Lyric-Vorschlag `try? e`
fun limitOrDefault(args: List<String>): Int =
    runCatching { parseArgs(args) }.getOrNull()?.limit ?: 1

fun main() {
    runB(listOf("-v", "x")); runB(listOf())
    println("default ${limitOrDefault(listOf())}")
}
// Ausgabe:
// B: limit 2 verbose true
// usage: need at least one argument
// default 1

File: docs/Befunde_und_Verbesserungen/prototypes/04-try-catch-definite-assignment/vergleich.cs
// 04 Vergleich (Stufe a): C# — Definite Assignment (ECMA-334 §9.4.4.x): nach einem
// try-Statement ist v definitely assigned, wenn v am Ende des try-Blocks UND am Ende
// jedes catch-Blocks definitely assigned ist — und ein catch-Block, der mit `return`
// endet, hat einen unerreichbaren Endpunkt, an dem ALLES als zugewiesen gilt.
using System;

class UsageError : Exception { public UsageError(string what) : base("usage: " + what) { } }
record struct Options(int Limit, bool Verbose);

static class Program
{
    static Options ParseArgs(string[] args)
    {
        if (args.Length == 0) throw new UsageError("need at least one argument");
        return new Options(args.Length, args[0] == "-v");
    }

    static int RunA(string[] args)
    {
        Options opts;
        try { opts = ParseArgs(args); }
        catch (UsageError e) { Console.WriteLine(e.Message); return 2; }
        Console.WriteLine($"A: limit {opts.Limit} verbose {opts.Verbose}");   // ok: opts ist zugewiesen
        return 0;
    }

    static int Main() => RunA(new[] { "-v", "x" }) + RunA(Array.Empty<string>());
}
// Ausgabe:
// A: limit 2 verbose True
// usage: need at least one argument
